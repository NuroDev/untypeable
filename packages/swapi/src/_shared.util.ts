import { beforeAll } from "vitest";
import { createTypeLevelClient } from "untypeable";
import { fetch } from "undici";
import { join } from "node:path";

import type { SwapiRouter } from ".";

export function useTestClient() {
  const client = createTypeLevelClient<SwapiRouter>(
    async (path, input = {}) => {
      const url = new URL(join("/api", path), "https://swapi.dev/");
      Object.entries(input).forEach(([key, value]) =>
        url.searchParams.append(key, value as string)
      );

      const response = await fetch(url.href, {
        method: "GET",
        headers: {
          "Content-Type": "application/json",
        },
      });

      return await response.json();
    }
  );

  beforeAll(() => {
    if (!client) throw "Failed to initialise untypeable client instance.";
  });

  return client;
}
