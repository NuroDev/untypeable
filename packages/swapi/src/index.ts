import { initUntypeable } from "untypeable";

import type { Film, FilmParams, Films } from "./films.types";
import type { People, Person, PersonParams } from "./people.types";
import type { Root } from "./root.types";
import type { AllSpecies, Species, SpeciesParams } from "./species.types";
import type { Starship, StarshipParams, Starships } from "./starships.types";
import type { Vehicle, VehicleParams, Vehicles } from "./vehicles.types";

const u = initUntypeable();

const router = u.router({
  "/": u.output<Root>(),
  /** Get all the film resources */
  "/films": u.output<Films>(),
  /** Get a specific film resource */
  "/films/:id": u.input<FilmParams>().output<Film>(),
  /** Get all the people resources */
  "/people": u.output<People>(),
  /** Get a specific people resource */
  "/people/:id": u.input<PersonParams>().output<Person>(),
  /** Get all the species resources */
  "/species": u.output<AllSpecies>(),
  /** Get a specific species resource */
  "/species/:id": u.input<SpeciesParams>().output<Species>(),
  /** Get all the starship resources */
  "/starships": u.output<Starships>(),
  /** Get a specific starship resource */
  "/starships/:id": u.input<StarshipParams>().output<Starship>(),
  /** Get all the vehicle resources */
  "/vehicles": u.output<Vehicles>(),
  /** Get a specific vehicle resource */
  "/vehicles/:id": u.input<VehicleParams>().output<Vehicle>(),
});

export type SwapiRouter = typeof router;
