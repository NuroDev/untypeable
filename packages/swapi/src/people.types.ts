import type { z } from "zod";

import type {
  PeopleSchema,
  PersonParamsSchema,
  PersonSchema,
} from "./people.validators";

/**
 * @docs https://swapi.dev/documentation#people
 */
export type Person = z.infer<typeof PersonSchema>;

/**
 * @docs https://swapi.dev/documentation#people
 */
export type People = z.infer<typeof PeopleSchema>;

export type PersonParams = z.infer<typeof PersonParamsSchema>;
