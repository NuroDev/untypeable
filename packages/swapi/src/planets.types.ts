import type { z } from "zod";

import type {
  PlanetParamsSchema,
  PlanetSchema,
  PlanetsSchema,
} from "./planets.validators";

/**
 * @docs https://swapi.dev/documentation#planets
 */
export type Planet = z.infer<typeof PlanetSchema>;

/**
 * @docs https://swapi.dev/documentation#planets
 */
export type Planets = z.infer<typeof PlanetsSchema>;

export type PlanetParams = z.infer<typeof PlanetParamsSchema>;
