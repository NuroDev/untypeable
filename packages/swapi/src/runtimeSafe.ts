import { initUntypeable } from "untypeable";

import {
  FilmSchema,
  FilmParamsSchema,
  FilmsSchema,
} from "./films/films.validators";
import {
  PeopleSchema,
  PersonSchema,
  PersonParamsSchema,
} from "./people/people.validators";
import {
  PlanetSchema,
  PlanetParamsSchema,
  PlanetsSchema,
} from "./planets/planets.validators";
import { RootSchema } from "./root/root.validators";
import {
  AllSpeciesSchema,
  SpeciesSchema,
  SpeciesParamsSchema,
} from "./species/species.validators";
import {
  StarshipSchema,
  StarshipParamsSchema,
  StarshipsSchema,
} from "./starships/starships.validators";
import {
  VehicleSchema,
  VehicleParamsSchema,
  VehiclesSchema,
} from "./vehicles/vehicles.validators";

const u = initUntypeable();

const router = u.router({
  "/": u.output(RootSchema),
  /** Get all the film resources */
  "/films": u.output(FilmsSchema),
  /** Get a specific film resource */
  "/films/:id": u.input(FilmParamsSchema).output(FilmSchema),
  /** Get all the people resources */
  "/people": u.output(PeopleSchema),
  /** Get a specific people resource */
  "/people/:id": u.input(PersonParamsSchema).output(PersonSchema),
  /** Get all the planets resources */
  "/planets": u.output(PlanetsSchema),
  /** Get a specific planets resource */
  "/planets/:id": u.input(PlanetParamsSchema).output(PlanetSchema),
  /** Get all the species resources */
  "/species": u.output(AllSpeciesSchema),
  /** Get a specific species resource */
  "/species/:id": u.input(SpeciesParamsSchema).output(SpeciesSchema),
  /** Get all the starship resources */
  "/starships": u.output(StarshipsSchema),
  /** Get a specific starship resource */
  "/starships/:id": u.input(StarshipParamsSchema).output(StarshipSchema),
  /** Get all the vehicle resources */
  "/vehicles": u.output(VehiclesSchema),
  /** Get a specific vehicle resource */
  "/vehicles/:id": u.input(VehicleParamsSchema).output(VehicleSchema),
});

export type SwapiRouter = typeof router;
