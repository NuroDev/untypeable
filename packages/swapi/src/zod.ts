export * from "./films.validators";
export * from "./people.validators";
export * from "./planets.validators";
export * from "./root.validators";
export * from "./species.validators";
export * from "./starships.validators";
export * from "./vehicles.validators";
