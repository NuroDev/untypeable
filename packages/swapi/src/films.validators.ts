import { z } from "zod";
import { defineSWAPIPaginatedSchema } from "./_shared/_shared.validators";

/**
 * @docs https://swapi.dev/documentation#films
 */
export const FilmSchema = z.object({
  /**
   * An array of people resource URLs that are in this film
   */
  characters: z.array(z.string().url()),
  /**
   * The ISO 8601 date format of the time that this resource was created */
  created: z.string().datetime(),
  /**
   * The name of the director of this film */
  director: z.string(),
  /**
   * The ISO 8601 date format of the time that this resource was edited */
  edited: z.string().datetime(),
  /**
   * The episode number of this film */
  episode_id: z.number(),
  /**
   * The opening paragraphs at the beginning of this film */
  opening_crawl: z.string(),
  /**
   * An array of planet resource URLs that are in this film */
  planets: z.array(z.string().url()),
  /**
   * The name(s) of the producer(s) of this film. Comma separated */
  producer: z.string(),
  /**
   * The ISO 8601 date format of film release at original creator country */
  release_date: z.string(),
  /**
   * An array of species resource URLs that are in this film */
  species: z.array(z.string().url()),
  /**
   * An array of starship resource URLs that are in this film */
  starships: z.array(z.string().url()),
  /**
   * The title of this film */
  title: z.string(),
  /**
   * The hypermedia URL of this resource */
  url: z.string().url(),
  /**
   * An array of vehicle resource URLs that are in this film */
  vehicles: z.array(z.string().url()),
});

/**
 * @docs https://swapi.dev/documentation#films
 */
export const FilmsSchema = defineSWAPIPaginatedSchema(FilmSchema);

export const FilmParamsSchema = z.object({
  id: z.number(),
});
