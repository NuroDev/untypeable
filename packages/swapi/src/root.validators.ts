import { z } from "zod";

/**
 * @docs https://swapi.dev/documentation#root
 */
export const RootSchema = z.object({
  /** The URL root for Film resources */
  films: z.literal("https://swapi.dev/api/films/"),
  /** The URL root for People resources */
  people: z.literal("https://swapi.dev/api/people/"),
  /** The URL root for Planet resources */
  planets: z.literal("https://swapi.dev/api/planets/"),
  /** The URL root for Species resources */
  species: z.literal("https://swapi.dev/api/species/"),
  /** The URL root for Starships resources */
  starships: z.literal("https://swapi.dev/api/starships/"),
  /** The URL root for Vehicles resources */
  vehicles: z.literal("https://swapi.dev/api/vehicles/"),
});
