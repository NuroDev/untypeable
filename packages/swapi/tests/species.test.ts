import { describe, it, expect } from "vitest";

import { AllSpeciesSchema, SpeciesSchema } from "../src/zod";
import { useTestClient } from "./_client";

describe.concurrent("SWAPI - Species", () => {
  const client = useTestClient();

  it("GET - /species", async () => {
    const allSpecies = await client("/species");

    expect(allSpecies).toBeDefined();
    expect(allSpecies.count).toBeDefined();
    expect(allSpecies.next).toBeDefined();
    expect(allSpecies.previous).toBeDefined();
    expect(Array.isArray(allSpecies.results)).toBe(true);
    expect(allSpecies.results.length).toBeGreaterThan(0);
    expect(allSpecies.results.at(0)).toBeDefined();

    expect(AllSpeciesSchema.safeParse(allSpecies).success).toBe(true);
  });

  it("GET - /species/:id", async () => {
    const species = await client("/species/:id", {
      id: 3,
    });

    expect(species).toBeDefined();
    expect(species).toMatchSnapshot({
      average_height: "210",
      average_lifespan: "400",
      classification: "mammal",
      created: "2014-12-10T16:44:31.486000Z",
      designation: "sentient",
      edited: "2014-12-20T21:36:42.142000Z",
      eye_colors: "blue, green, yellow, brown, golden, red",
      films: [
        "https://swapi.dev/api/films/1/",
        "https://swapi.dev/api/films/2/",
        "https://swapi.dev/api/films/3/",
        "https://swapi.dev/api/films/6/",
      ],
      hair_colors: "black, brown",
      homeworld: "https://swapi.dev/api/planets/14/",
      language: "Shyriiwook",
      name: "Wookie",
      people: [
        "https://swapi.dev/api/people/13/",
        "https://swapi.dev/api/people/80/",
      ],
      skin_colors: "gray",
      url: "https://swapi.dev/api/species/3/",
    });

    expect(SpeciesSchema.safeParse(species).success).toBe(true);
  });
});
