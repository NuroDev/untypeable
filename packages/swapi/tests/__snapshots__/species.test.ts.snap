// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SWAPI - Species > GET - /species/:id 1`] = `
{
  "average_height": "210",
  "average_lifespan": "400",
  "classification": "mammal",
  "created": "2014-12-10T16:44:31.486000Z",
  "designation": "sentient",
  "edited": "2014-12-20T21:36:42.142000Z",
  "eye_colors": "blue, green, yellow, brown, golden, red",
  "films": [
    "https://swapi.dev/api/films/1/",
    "https://swapi.dev/api/films/2/",
    "https://swapi.dev/api/films/3/",
    "https://swapi.dev/api/films/6/",
  ],
  "hair_colors": "black, brown",
  "homeworld": "https://swapi.dev/api/planets/14/",
  "language": "Shyriiwook",
  "name": "Wookie",
  "people": [
    "https://swapi.dev/api/people/13/",
    "https://swapi.dev/api/people/80/",
  ],
  "skin_colors": "gray",
  "url": "https://swapi.dev/api/species/3/",
}
`;
